finality-aleph/src/sync/substrate/justification.rs

```diff
@@ -0,0 +1,166 @@
+use std::fmt::{Display, Error as FmtError, Formatter};
+
+use aleph_primitives::{BlockNumber, SessionAuthorityData};
+use codec::Encode;
+use log::warn;
+use sp_runtime::{
+    traits::{Block, Header as SubstrateHeader},
+    RuntimeAppPublic,
+};
+
+use crate::{
+    crypto::AuthorityVerifier,
+    justification::{AlephJustification, Verifier as LegacyVerifier},
+    session::session_id_from_num,
+    session_map::ReadOnlySessionMap,
+    sync::{Justification as JustificationT, Verifier},
+    AuthorityId, SessionPeriod,
+};
+
+/// A justification, including the related header.
+#[derive(Clone)]
+pub struct Justification<H: SubstrateHeader<Number = BlockNumber>> {
+    header: H,
+    raw_justification: AlephJustification,
+}
+
+impl<H: SubstrateHeader<Number = BlockNumber>> JustificationT for Justification<H> {
+    type Header = H;
+    type Unverified = Self;
+
+    fn header(&self) -> &Self::Header {
+        &self.header
+    }
+
+    fn into_unverified(self) -> Self::Unverified {
+        self
+    }
+}
+
+/// A justification verifier within a single session.
+pub struct SessionVerifier {
+    authority_verifier: AuthorityVerifier,
+    emergency_signer: Option<AuthorityId>,
+}
+
+impl From<SessionAuthorityData> for SessionVerifier {
+    fn from(authority_data: SessionAuthorityData) -> Self {
+        SessionVerifier {
+            authority_verifier: AuthorityVerifier::new(authority_data.authorities().to_vec()),
+            emergency_signer: authority_data.emergency_finalizer().clone(),
```

Comment:
Also this clone should probably not happen, as we move `SessionAuthorityData`. That said I'm not sure how to get authorities and signer out of `SessionAuthorityData` ://